### PR TITLE
[HZ-331] Preliminary preparation for IMap index scan in Jet SQL

### PR DESCRIPTION
**Pull request by Fly-Style**
The scope of this PR was changed.
We are decided to remove `OnHeapMapIndexScanP` (the processor with direct access to `indexStore`), but in the same time to leave the rest of the code. 
This code a good prerequisite for future migration-tolerant index scan processor which will use operation to scan the index.

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Request reviewers if possible

### CODE REVIEW

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/IMapSqlConnector.java [hunk 1]
```diff
@@ -136,6 +142,30 @@ public Table createTable(
         );
     }
 
+    @Nonnull
+    @Override
+    public VertexWithInputConfig nestedLoopReader(
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/IMapSqlConnector.java:147]: Can we stick to interface method order?

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/IMapSqlConnector.java [hunk 2]
```diff
@@ -155,31 +185,44 @@ public Vertex fullScanReader(
                 filter
         );
 
-        return dag.newUniqueVertex(table.toString(), OnHeapMapScanP.onHeapMapScanP(mapScanMetadata));
+        return dag.newUniqueVertex(table.getMapName(), OnHeapMapScanP.onHeapMapScanP(mapScanMetadata));
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/IMapSqlConnector.java:188]: What's wrong with `table.toString()`?

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/IMapSqlConnector.java [hunk 3]
```diff
@@ -155,31 +185,44 @@ public Vertex fullScanReader(
...
-    @Nonnull
-    @Override
-    public VertexWithInputConfig nestedLoopReader(
+    public Vertex indexScanReader(
             @Nonnull DAG dag,
             @Nonnull Table table0,
-            @Nullable Expression<Boolean> predicate,
-            @Nonnull List<Expression<?>> projections,
-            @Nonnull JetJoinInfo joinInfo
+            @Nonnull IMapIndexScanPhysicalRel rel,
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/IMapSqlConnector.java:194]: We must avoid exposing `Calcite` related classes to possibly-public API.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/FullScanPhysicalRule.java [hunk 4]
```diff
@@ -46,4 +57,38 @@ public RelNode convert(RelNode rel) {
                 logicalScan.getTable()
         );
     }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/FullScanPhysicalRule.java:62]: If you need to override `onMatch()` most probably the rule should simply extend `RelOptRule` not `ConverterRule`.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/JetPlan.java [hunk 5]
```diff
@@ -38,7 +38,7 @@
 import java.util.List;
 import java.util.Set;
 
-abstract class JetPlan extends SqlPlan {
+public abstract class JetPlan extends SqlPlan {
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/JetPlan.java:41]: Why these have to public now?

File: hazelcast/src/main/java/com/hazelcast/sql/impl/SqlDataSerializerHook.java [hunk 6]
```diff
@@ -201,10 +202,9 @@
     public static final int EXPRESSION_EXTRACT = 72;
 
     public static final int MAP_SCAN_METADATA = 73;
-//    Reserved
-//    public static final int MAP_INDEX_SCAN_METADATA = 74;
+    public static final int MAP_INDEX_SCAN_METADATA = 74;
 
-    public static final int LEN = MAP_SCAN_METADATA + 1;
+    public static final int LEN = MAP_INDEX_SCAN_METADATA + 1;
```
Comment (gierlachg) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/SqlDataSerializerHook.java:207]: This is confusing, some parts of Map/Map Index scanning serialization are defined here, other in `JetSqlSerializerHook`.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/JetPlan.java (hunk 5 above)
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/JetPlan.java:41]: For testing purposes.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/SqlDataSerializerHook.java (hunk 6 above)
Comment (Fly-Style) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/SqlDataSerializerHook.java:207]: Processors serialization defined in `JetSqlSerializerHook`, everything else in `SqlSerializerHook`

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/IMapSqlConnector.java (hunk 1 above)
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/IMapSqlConnector.java:147]: Sure, I am sorry.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/IMapSqlConnector.java (hunk 2 above)
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/IMapSqlConnector.java:188]: More clear description in debug info; easier to test.

Comment (gierlachg) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/IMapSqlConnector.java:188]: Maybe `toString()` defined in this class is what you need? (Consider adding it to `PartitionedMapTable`)

**Comment by Fly-Style** [comment]
run-lab-run

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/IMapIndexScanPhysicalRel.java [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    @Override
+    public RelWriter explainTerms(RelWriter pw) {
+        return super.explainTerms(pw)
+                .item("index", index.getName())
+                .item("indexExp", indexExp)
+                .item("remainderExp", remainderExp);
+    }
+
+    @Override
+    public double estimateRowCount(RelMetadataQuery mq) {
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/IMapIndexScanPhysicalRel.java:164]: I can see `MapIndexScanPhysicalRel` overrides `computeSelfCost()`, we don't need that?

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateDagVisitor.java [hunk 8]
```diff
@@ -122,6 +123,25 @@ public Vertex onFullScan(FullScanPhysicalRel rel) {
         return jetSqlConnector.fullScanReader(dag, table, rel.filter(parameterMetadata), rel.projection(parameterMetadata));
     }
 
+    public Vertex onIndexScan(IMapIndexScanPhysicalRel rel) {
+        Table table = rel.getTable().unwrap(HazelcastTable.class).getTarget();
+        collectObjectKeys(table);
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateDagVisitor.java:128]: Let's create a test that proves that plan cache works with `IMap` indices.

**Comment by gierlachg** [comment]
I think we should add indices to created `PartitionedMapTable` in `IMapSqlConnector.createTable()`.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapIndexScanP.java [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+    @Override
+    public boolean complete() {
+        return emitFromTraverser(traverser);
+    }
+
+    static class IMapIndexTraverser implements Traverser<Object[]> {
+        private final MapContainer mapContainer;
+        private final List<Expression<?>> projections;
+        private final Expression<Boolean> filter;
+        private PartitionIdSet partitions;
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapIndexScanP.java:117]: It is not used.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapIndexScanP.java [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+        private final MapScanRow row;
+        private final SimpleExpressionEvalContext ctx;
+
+        private InternalIndex index;
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapIndexScanP.java:121]: Can be converted to local variable.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapIndexScanP.java [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+
+        private final String indexName;
+        private final IndexFilter indexFilter;
+        private final List<QueryDataType> converterTypes;
+        private final int componentCount;
+        // The collation of index entries
+        private List<Boolean> ascs;
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapIndexScanP.java:128]: Can be final.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapIndexScanP.java [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+
+        /**
+         * Stamp to ensure that indexed partitions are stable throughout query execution.
+         */
+        private Long partitionStamp;
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapIndexScanP.java:133]: It is not used.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapIndexScanP.java [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+
+        private final KeyValueIterator iteratorExec;
+
+        IMapIndexTraverser(
+                @Nonnull final NodeEngine nodeEngine,
+                @Nonnull final MapIndexScanMetadata indexScanMetadata,
+                @Nonnull final PartitionIdSet partitions,
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapIndexScanP.java:140]: It is not used.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapIndexScanP.java [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+                    indexScanMetadata.getFieldPaths(),
+                    indexScanMetadata.getFieldTypes(),
+                    mapContainer.getExtractors(),
+                    ctx.getSerializationService()
+            );
+        }
+
+        @Override
+        public Object[] next() {
+            while (iteratorExec.tryAdvance()) {
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapIndexScanP.java:169]: We should  limit the number of rows scanned in one call to not break the cooperative contract, similarly to `OnHeapMapScanP`'s traverser.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapIndexScanP.java [hunk 15]
```diff
@@ -0,0 +1,390 @@
...
+
+            if (index == null) {
+                throw QueryException.error(
+                        SqlErrorCode.INDEX_INVALID,
+                        "Cannot use the index \"" + indexName + "\" of the IMap \""
+                                + mapContainer.getName() + "\" because it doesn't exist"
+                ).markInvalidate();
+            }
+
+            // Make sure that required partitions are indexed
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapIndexScanP.java:206]: Is it a forgotten TODO?

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapIndexScanP.java [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+                    index,
+                    componentCount,
+                    indexFilter,
+                    ascs,
+                    converterTypes,
+                    ctx
+            );
+        }
+
+        private QueryException invalidIndexStamp() {
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapIndexScanP.java:224]: It is not used.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/SqlDataSerializerHook.java (hunk 6 above)
Comment (gierlachg) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/SqlDataSerializerHook.java:207]: Metadata classes are used solely by custom IMap processors. They could be moved to the same package, access modifiers adjusted and serialization configuration could be placed in single hook.

File: hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/opt/physical/index/JetIndexResolver.java [hunk 17]
```diff
@@ -0,0 +1,1445 @@
...
+import static java.util.Collections.singletonList;
+import static org.apache.calcite.rel.RelFieldCollation.Direction;
+import static org.apache.calcite.rel.RelFieldCollation.Direction.ASCENDING;
+import static org.apache.calcite.rel.RelFieldCollation.Direction.DESCENDING;
+
+/**
+ * Helper class to resolve indexes.
+ */
+@SuppressWarnings({"rawtypes", "checkstyle:MethodCount"})
+public final class JetIndexResolver {
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/opt/physical/index/JetIndexResolver.java:90]: It's an `IMap` specific utility. I would move it to `IMap` specific package.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/index/JetSqlIndexAbstractTest.java [hunk 18]
```diff
@@ -0,0 +1,709 @@
...
+import static com.hazelcast.sql.support.expressions.ExpressionTypes.INTEGER;
+import static com.hazelcast.sql.support.expressions.ExpressionTypes.LONG;
+import static com.hazelcast.sql.support.expressions.ExpressionTypes.SHORT;
+import static com.hazelcast.sql.support.expressions.ExpressionTypes.STRING;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+@SuppressWarnings({"unchecked", "rawtypes", "unused"})
+public abstract class JetSqlIndexAbstractTest extends SqlTestSupport {
```
Comment (gierlachg) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/index/JetSqlIndexAbstractTest.java:92]: This class tests specifically `IMap` behaviour. I would move it to `IMap` specific package.
Moreover, let's make sure Jet engine is used by this test by using a Jet engine feature that is not supported by IMDG engine.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/exec/scan/index/IndexFilter.java [hunk 19]
```diff
@@ -20,13 +20,14 @@
...
 
+import java.io.Serializable;
 import java.util.Iterator;
 
 /**
  * Index filter which is transferred over a wire.
  */
 @SuppressWarnings("rawtypes")
-public interface IndexFilter {
+public interface IndexFilter extends Serializable {
```
Comment (gierlachg) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/exec/scan/index/IndexFilter.java:30]: I believe it is not needed anymore.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/FullScanPhysicalRule.java [hunk 20]
```diff
@@ -46,4 +57,38 @@ public RelNode convert(RelNode rel) {
...
+        RelNode rel = call.rel(0);
+        if (rel.getTraitSet().contains(getInTrait())) {
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/FullScanPhysicalRule.java:64]: I'm pretty sure this condition is always `true`.

**Review by gierlachg — CHANGES_REQUESTED** [review]

**Comment by Fly-Style** [comment]
run-lab-run

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/FullScanPhysicalRule.java (hunk 4 above)
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/FullScanPhysicalRule.java:61]: IMO, no need to do that, it's only overcomplicate the implementation.

Comment (gierlachg) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/FullScanPhysicalRule.java:61]: It actually simplifies it, `RelOptRule` is more generic.

**Review by hakanmemisoglu — APPROVED** [review]
It looks good!
A minor comment: 
In our codebase, I mostly saw `descending` is used to specify order. It would be better to use it instead of `asc`. It is your decision.

**Pull request closed without merging**